File: viewer-web/core/solvers.js
// Solvers, ported from src/solvers.cpp: Adam (Kingma & Ba 2015 Algorithm 1) on the phase with
// wrapping to the torus and optional stepwise quantization by continuation (Wyrowski 1990 eq. 22,
// capture ramp linear per iteration after Skeren et al. 2002 eq. 11 by default, or Wyrowski's held table;
// as a projected step after 40 % of the budget); Gerchberg-Saxton with Wyrowski's two stages
// (operator X with the field zeroed outside the window, then X' with the amplitude free;
// eqs. 7-10) and his quantization cycles J = Q(P-1)+1 (section 3.C).
(function () {
  // Module factory. As a classic script it also registers its own source so that the page can
  // build a Web Worker from it (a worker script cannot be loaded from a file:// URL).
  function factory(root) {
  const isNode = typeof module !== 'undefined' && module.exports;
  const { wrapToPi } = isNode ? require('./util.js') : root.DOE_CORE;
  const { energyAndGrad, optimalScale } = isNode ? require('./energy.js') : root.DOE_CORE;
  const { projectToLevels, projectStepwise, wyrowskiEpsilon, linearEpsilon } = isNode ? require('./quantize.js') : root.DOE_CORE;

  class Adam {
    constructor(size, params) {
      this.p = Object.assign({ lr: 0.05, beta1: 0.9, beta2: 0.999, eps: 1e-8 }, params);
      this.m = new Float64Array(size); this.v = new Float64Array(size); this.t = 0;
    }
    step(x, g) {
      if (x.length !== this.m.length || g.length !== this.m.length) throw new Error('Adam.step: shape mismatch');
      ++this.t;
      const { lr, beta1, beta2, eps } = this.p, c1 = 1 - beta1 ** this.t, c2 = 1 - beta2 ** this.t;
      for (let k = 0; k < x.length; ++k) {
        const gk = g[k];
        this.m[k] = beta1 * this.m[k] + (1 - beta1) * gk;
        this.v[k] = beta2 * this.v[k] + (1 - beta2) * gk * gk;
        x[k] -= lr * (this.m[k] / c1) / (Math.sqrt(this.v[k] / c2) + eps);
      }
    }
  }
  const wrapPhase = (phi) => { for (let k = 0; k < phi.length; ++k) phi[k] = wrapToPi(phi[k]); };
  function fieldOf(phi, illum, prop) {
    const N = phi.length, re = new Float64Array(N), im = new Float64Array(N);
    for (let k = 0; k < N; ++k) { re[k] = illum[k] * Math.cos(phi[k]); im[k] = illum[k] * Math.sin(phi[k]); }
    return prop.forward({ re, im });
  }

  // params: { iters, lr, weights: {shape, efficiency}, adam: {beta1, beta2, eps}, quant: {levels, start, ramp, steps} }
  // progress(it, energy) -> false stops early.
  function optimize(phi0, illum, prop, b, mask, params, progress) {
    const P = Object.assign({ iters: 400, lr: 0.05 }, params);
    const weights = Object.assign({ shape: 1, efficiency: 0.3 }, P.weights);
    const qp = Object.assign({ levels: 0, start: 0.4, ramp: 'linear', steps: 10 }, P.quant);
    const r = { phi: Float64Array.from(phi0), history: [], shapeHistory: [], efficiencyHistory: [] };
    const adam = new Adam(phi0.length, Object.assign({}, P.adam, { lr: P.lr }));
    const q = qp.levels, it0 = q > 0 ? Math.floor(qp.start * P.iters) : P.iters;
    const remaining = Math.max(P.iters - it0, 1), steps = Math.max(qp.steps, 1);
    for (let it = 0; it < P.iters; ++it) {
      const e = energyAndGrad(r.phi, illum, prop, b, mask, weights);
      r.history.push(e.energy); r.shapeHistory.push(e.shape); r.efficiencyHistory.push(e.efficiency);
      if (progress && progress(it, e.energy) === false) break;
      adam.step(r.phi, e.grad);
      wrapPhase(r.phi);
      if (q > 0 && it >= it0) {
        const eps = qp.ramp === 'linear' ? linearEpsilon(it - it0 + 1, remaining)
                                         : wyrowskiEpsilon(Math.min(steps, 1 + Math.floor((it - it0) * steps / remaining)), steps);
        projectStepwise(r.phi, q, eps);
      }
    }
    if (q > 0) projectToLevels(r.phi, q);
    r.field = fieldOf(r.phi, illum, prop);
    return r;
  }

  // params: { iters, phaseOnlyIters, quant: {levels, ramp, steps, cyclesPerStep} }
  function gerchbergSaxton(phi0, illum, prop, b, mask, params) {
    const P = Object.assign({ iters: 40, phaseOnlyIters: 20 }, params);
    const qp = Object.assign({ levels: 0, ramp: 'linear', steps: 10, cyclesPerStep: 5 }, P.quant);
    const q = qp.levels, steps = Math.max(qp.steps, 1), Q = Math.max(qp.cyclesPerStep, 1);
    const quantCycles = q > 0 ? Q * (steps - 1) + 1 : 0, iters = P.iters + quantCycles;
    const N = phi0.length;
    const r = { phi: Float64Array.from(phi0), history: [], shapeHistory: [], efficiencyHistory: [], residualHistory: [] };
    const vAbs = new Float64Array(N), nre = new Float64Array(N), nim = new Float64Array(N);
    for (let it = 0; it < iters; ++it) {
      const e = energyAndGrad(r.phi, illum, prop, b, mask, { shape: 1, efficiency: 0.3 });
      r.history.push(e.energy); r.shapeHistory.push(e.shape); r.efficiencyHistory.push(e.efficiency);
      const v = e.field;
      for (let k = 0; k < N; ++k) vAbs[k] = Math.hypot(v.re[k], v.im[k]);
      const s = optimalScale(vAbs, b, mask);
      const freeOutside = it >= P.phaseOnlyIters || it >= P.iters;
      let res = 0;
      for (let k = 0; k < N; ++k) {
        const f = vAbs[k];
        if (mask[k] > 0) { const d = f - s * b[k]; res += d * d; } else if (!freeOutside) res += f * f;
      }
      r.residualHistory.push(res);
      for (let k = 0; k < N; ++k) {
        const a = Math.max(vAbs[k], 1e-12), inWindow = mask[k] > 0;
        const amp = inWindow ? s * b[k] : (freeOutside ? vAbs[k] : 0);
        nre[k] = amp * v.re[k] / a; nim[k] = amp * v.im[k] / a;
      }
      const back = prop.adjoint({ re: nre, im: nim });
      for (let k = 0; k < N; ++k) r.phi[k] = wrapToPi(Math.atan2(back.im[k], back.re[k]));
      if (q > 0 && it >= P.iters) {
        const j = it - P.iters + 1;
        const eps = qp.ramp === 'linear' ? linearEpsilon(j, quantCycles)
                                         : wyrowskiEpsilon(Math.min(steps, 1 + Math.floor((j - 1) / Q)), steps);
        projectStepwise(r.phi, q, eps);
      }
    }
    if (q > 0) projectToLevels(r.phi, q);
    wrapPhase(r.phi);
    r.field = fieldOf(r.phi, illum, prop);
    return r;
  }

  const api = { Adam, optimize, gerchbergSaxton, fieldOf };
  if (isNode) module.exports = api; else { root.DOE_CORE = root.DOE_CORE || {}; Object.assign(root.DOE_CORE, api); }
  }
  const root = typeof globalThis !== 'undefined' ? globalThis : this;
  factory(root);
  if (!(typeof module !== 'undefined' && module.exports)) (root.DOE_MODULE_SOURCES = root.DOE_MODULE_SOURCES || []).push(factory.toString());
})();

File: src/solvers.cpp
#include "doe/solvers.hpp"

#include "doe/phase.hpp"

#include <cmath>
#include <complex>
#include <numbers>
#include <stdexcept>

namespace doe {

template <class T>
Adam<T>::Adam(std::size_t rows, std::size_t cols, AdamParams params)
    : p_(params), m_(rows, cols, 0.0), v_(rows, cols, 0.0) {}

template <class T>
void Adam<T>::step(Array2<T>& x, const Array2<T>& g) {
    if (x.rows != m_.rows || x.cols != m_.cols || g.rows != m_.rows || g.cols != m_.cols)
        throw std::invalid_argument("Adam::step: shape mismatch");
    ++t_;
    const double c1 = 1.0 - std::pow(p_.beta1, t_), c2 = 1.0 - std::pow(p_.beta2, t_);
    for (std::size_t k = 0; k < m_.data.size(); ++k) {
        const double gk = double(g.data[k]);
        m_.data[k] = p_.beta1 * m_.data[k] + (1.0 - p_.beta1) * gk;
        v_.data[k] = p_.beta2 * v_.data[k] + (1.0 - p_.beta2) * gk * gk;
        const double mhat = m_.data[k] / c1, vhat = v_.data[k] / c2;
        x.data[k] = T(double(x.data[k]) - p_.lr * mhat / (std::sqrt(vhat) + p_.eps));
    }
}

template <class T>
void wrap_phase(Array2<T>& phi) {
    for (auto& x : phi.data) x = T(wrap_to_pi(double(x)));
}

template <class T>
Result<T> gerchberg_saxton(const Array2<T>& phi0, const Array2<T>& illum, const AngularSpectrum<T>& prop,
                           const Array2<T>& b, const Array2<T>& mask, const GsParams& params) {
    const int q = params.quant.levels;
    // Wyrowski §3.C: Q cycles for each of the steps p = 1..P-1, one cycle for p = P.
    const int P = std::max(params.quant.steps, 1), Q = std::max(params.quant.cycles_per_step, 1);
    const int quant_cycles = q > 0 ? Q * (P - 1) + 1 : 0;
    const int iters = params.iters + quant_cycles;
    const std::size_t N = phi0.data.size();
    Result<T> r;
    r.phi = phi0;
    Field<T> u(phi0.rows, phi0.cols), v_new(phi0.rows, phi0.cols);
    Array2<T> v_abs(phi0.rows, phi0.cols);
    for (int it = 0; it < iters; ++it) {
        // Forward: v = A(illum e^{i phi}); log the normalized energy for comparison with Adam.
        auto e = energy_and_grad<T>(r.phi, illum, prop, b, mask, params.weights);
        r.history.push_back(e.energy);
        r.shape_history.push_back(e.shape);
        r.efficiency_history.push_back(e.efficiency);
        const Field<T>& v = e.field;
        for (std::size_t k = 0; k < N; ++k) v_abs.data[k] = T(std::abs(std::complex<double>(v.data[k])));
        const double s = optimal_scale(v_abs, b, mask);

        // Fienup's projection residual: squared distance of v to the constraint
        // set of this cycle. For X that set is {c b on W, 0 outside}, so the
        // outside energy counts; for X' only the window does. Non-increasing
        // within a stage when A is unitary (error reduction, Fienup 1982).
        const bool free_outside = it >= params.phase_only_iters || it >= params.iters;
        double res = 0.0;
        for (std::size_t k = 0; k < N; ++k) {
            const double f = double(v_abs.data[k]);
            if (mask.data[k] > 0) {
                const double d = f - s * double(b.data[k]);
                res += d * d;
            } else if (!free_outside) {
                res += f * f;
            }
        }
        r.residual_history.push_back(res);

        // (i) target-plane projection (Wyrowski eqs. 7 / 10): modulus c_j b on the
        // window, phase kept; outside the window zero (X, first stage) or
        // untouched (X', amplitude freedom).
        for (std::size_t k = 0; k < N; ++k) {
            const std::complex<double> vk(v.data[k]);
            const double a = std::max(double(v_abs.data[k]), 1e-12);
            const std::complex<double> phase = vk / a;
            const bool in_window = mask.data[k] > 0;
            const double amp = in_window ? s * double(b.data[k]) : (free_outside ? double(v_abs.data[k]) : 0.0);
            v_new.data[k] = std::complex<T>(amp * phase);
        }
        // (ii) DOE-plane projection (Wyrowski eq. 8): keep only the phase; |u| = illum is re-imposed by the forward model
        Field<T> back = prop.adjoint(v_new);
        for (std::size_t k = 0; k < N; ++k) r.phi.data[k] = T(wrap_to_pi(std::arg(std::complex<double>(back.data[k]))));
        // Quantization stage: Q_Z^(p) (Wyrowski eq. 22) replaces the unit-modulus
        // operator. Table ramp: step p advances every Q cycles, the last step is
        // direct. Linear ramp (Skeren et al. 2002 eq. 11): every one of the J
        // cycles is a step, epsilon = j / J.
        if (q > 0 && it >= params.iters) {
            const int j = it - params.iters + 1;
            const double eps = params.quant.ramp == QuantRamp::linear ? capture_fraction(QuantRamp::linear, j, quant_cycles)
                                                                       : capture_fraction(QuantRamp::table, std::min(P, 1 + (j - 1) / Q), P);
            project_stepwise(r.phi, q, eps);
        }
    }
    if (q > 0) project_to_levels(r.phi, q);
    wrap_phase(r.phi);
    for (std::size_t k = 0; k < N; ++k) u.data[k] = std::polar(illum.data[k], r.phi.data[k]);
    r.field = prop.forward(u);
    return r;
}

template <class T>
Result<T> optimize(const Array2<T>& phi0, const Array2<T>& illum, const AngularSpectrum<T>& prop,
                   const Array2<T>& b, const Array2<T>& mask, const OptimizeParams& params,
                   const ProgressFn& progress) {
    Result<T> r;
    r.phi = phi0;
    AdamParams ap = params.adam;
    ap.lr = params.lr;
    Adam<T> adam(phi0.rows, phi0.cols, ap);
    const QuantParams& qp = params.quant;
    const int q = qp.levels;
    const int it0 = q > 0 ? static_cast<int>(std::floor(qp.start * params.iters)) : params.iters;
    const int remaining = std::max(params.iters - it0, 1);
    const int P = std::max(qp.steps, 1);
    for (int it = 0; it < params.iters; ++it) {
        const bool quantizing = q > 0 && it >= it0;
        if (quantizing && qp.method == QuantMethod::choi) {
            // Choi eq. (5): forward with the hard quantizer, backward through the relaxation.
            Array2<T> phi_q = r.phi;
            project_to_levels(phi_q, q);
            auto e = energy_and_grad<T>(phi_q, illum, prop, b, mask, params.weights);
            r.history.push_back(e.energy);
            r.shape_history.push_back(e.shape);
            r.efficiency_history.push_back(e.efficiency);
            if (progress && !progress(it, e.energy)) break;
            const double frac = remaining > 1 ? double(it - it0) / double(remaining - 1) : 1.0;
            // Supplement S2.3: score scale 300 -> 1000, tau = tau0 exp(-c frac), w from the level spacing.
            const double w = qp.choi_w > 0 ? qp.choi_w : 4.0 * q / (2.0 * std::numbers::pi);
            const GumbelSoftmaxQuantizer Q{q, w, qp.choi_tau0 * std::exp(-qp.choi_c * frac),
                                           qp.choi_gain0 + (qp.choi_gain1 - qp.choi_gain0) * frac};
            const Array2<double> noise = Q.sample_noise(r.phi.data.size(), qp.choi_seed + 7919ULL * static_cast<unsigned long long>(it + 1));
            Array2<T> g = surrogate_gradient(e.grad, r.phi, Q, noise);
            adam.step(r.phi, g);
            wrap_phase(r.phi);
            continue;
        }
        auto e = energy_and_grad<T>(r.phi, illum, prop, b, mask, params.weights);
        r.history.push_back(e.energy);
        r.shape_history.push_back(e.shape);
        r.efficiency_history.push_back(e.efficiency);
        if (progress && !progress(it, e.energy)) break;
        adam.step(r.phi, e.grad);
        wrap_phase(r.phi);  // stay on the torus
        if (quantizing) {
            // Wyrowski eq. (22) as a projected gradient step. Linear ramp (Skeren
            // et al. 2002 eq. 11): every iteration of the stage is a step, epsilon
            // = (it - it0 + 1) / remaining. Table ramp: step p of P grows with the budget.
            const double eps = qp.ramp == QuantRamp::linear
                                   ? capture_fraction(QuantRamp::linear, it - it0 + 1, remaining)
                                   : capture_fraction(QuantRamp::table, std::min(P, 1 + (it - it0) * P / remaining), P);
            project_stepwise(r.phi, q, eps);
        }
    }
    if (q > 0) project_to_levels(r.phi, q);
    Field<T> u(phi0.rows, phi0.cols);
    for (std::size_t k = 0; k < u.data.size(); ++k) u.data[k] = std::polar(illum.data[k], r.phi.data[k]);
    r.field = prop.forward(u);
    return r;
}

template class Adam<float>;
template class Adam<double>;
template void wrap_phase<float>(Array2<float>&);
template void wrap_phase<double>(Array2<double>&);
template Result<float> gerchberg_saxton<float>(const Array2<float>&, const Array2<float>&, const AngularSpectrum<float>&,
                                               const Array2<float>&, const Array2<float>&, const GsParams&);
template Result<double> gerchberg_saxton<double>(const Array2<double>&, const Array2<double>&, const AngularSpectrum<double>&,
                                                 const Array2<double>&, const Array2<double>&, const GsParams&);
template Result<float> optimize<float>(const Array2<float>&, const Array2<float>&, const AngularSpectrum<float>&,
                                       const Array2<float>&, const Array2<float>&, const OptimizeParams&, const ProgressFn&);
template Result<double> optimize<double>(const Array2<double>&, const Array2<double>&, const AngularSpectrum<double>&,
                                         const Array2<double>&, const Array2<double>&, const OptimizeParams&, const ProgressFn&);

}  // namespace doe
